File: src/audit/chain.ts
// Tamper-evident hash chain over the append-only audit log.
// ADDITIVE ONLY: we do not change the existing entry schema — we append two
// new fields (prevHash, hash) to every entry. Each entry's hash commits to the
// previous entry's hash, forming a chain that makes any edit detectable.

import { createHash } from 'crypto';
import { AuditEntry } from '../flows/types';

export const GENESIS_HASH = '0'.repeat(64);

// Stable key-sorted stringify so the hash is reproducible regardless of how the
// object was constructed (key insertion order). ALWAYS excludes prevHash/hash
// so the hash does not self-reference, even if the object still carries them.
export function stableStringifyEntry(entry: Omit<AuditEntry, 'prevHash' | 'hash'>): string {
  const clone: Record<string, unknown> = {};
  for (const key of Object.keys(entry).sort()) {
    if (key === 'prevHash' || key === 'hash') continue;
    const value = (entry as unknown as Record<string, unknown>)[key];
    if (value !== undefined) clone[key] = value;
  }
  return JSON.stringify(clone);
}

// The data actually committed by the hash: every field except the two hash fields.
function hashPayload(entry: AuditEntry): string {
  const { prevHash: _p, hash: _h, ...rest } = entry;
  return stableStringifyEntry(rest);
}

export function sha256(input: string): string {
  return createHash('sha256').update(input, 'utf8').digest('hex');
}

// Compute an entry's hash given its prevHash. prevHash must equal the prior
// entry's hash (or GENESIS_HASH for the first entry).
export function computeHash(prevHash: string, entry: Omit<AuditEntry, 'prevHash' | 'hash'>): string {
  return sha256(prevHash + stableStringifyEntry(entry));
}

export interface HashedEntry extends AuditEntry {
  prevHash: string;
  hash: string;
}

// Pure: given the log tail and new entry data, return the log with the new
// hashed entry appended. Does NOT mutate the input array.
export function appendAuditEntry(
  log: readonly HashedEntry[],
  newEntryData: Omit<AuditEntry, 'prevHash' | 'hash'>,
): HashedEntry[] {
  const prevHash = log.length > 0 ? log[log.length - 1].hash : GENESIS_HASH;
  const hash = computeHash(prevHash, newEntryData);
  return [...log, { ...newEntryData, prevHash, hash }];
}

export interface ChainVerification {
  valid: boolean;
  brokenAtIndex: number | null;
  entries: number;
}

// Recompute the entire chain from scratch. Returns the first broken index
// (or null if the whole chain verifies).
export function verifyChain(log: readonly AuditEntry[]): ChainVerification {
  for (let i = 0; i < log.length; i++) {
    const entry = log[i] as HashedEntry;
    const expectedPrev = i === 0 ? GENESIS_HASH : (log[i - 1] as HashedEntry).hash;

    if (entry.prevHash !== expectedPrev) {
      return { valid: false, brokenAtIndex: i, entries: log.length };
    }
    const recomputed = computeHash(entry.prevHash, entry);
    if (entry.hash !== recomputed) {
      return { valid: false, brokenAtIndex: i, entries: log.length };
    }
  }
  return { valid: true, brokenAtIndex: null, entries: log.length };
}

File: go/rz/chain.go
package rz

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
)

// Tamper-evident hash chain over the append-only audit log.
// ADDITIVE ONLY: we do not change the existing entry schema — we append two
// new fields (prevHash, hash) to every entry. Each entry's hash commits to the
// previous entry's hash, forming a chain that makes any edit detectable.

const GenesisHash = "0000000000000000000000000000000000000000000000000000000000000000"

// StablePayload is the stable key-sorted stringify of an entry, ALWAYS excluding
// prevHash/hash so the hash does not self-reference. Optional fields that were
// never set are omitted (mirroring JS). Sorted keys via encoding/json maps.
func StablePayload(e *AuditEntry) string {
	m := map[string]any{
		"eventId":      e.EventID,
		"timestamp":    e.Timestamp,
		"flow":         string(e.Flow),
		"reasonBucket": string(e.ReasonBucket),
		"ruleFired":    string(e.RuleFired),
		"decision":     string(e.Decision),
		"actor":        string(e.Actor),
		"outcome":      e.Outcome,
		"state":        string(e.State),
	}
	if e.Attempt != nil {
		m["attempt"] = *e.Attempt
	}
	if e.Amount != nil {
		m["amount"] = *e.Amount
	}
	if e.Currency != "" {
		m["currency"] = e.Currency
	}
	if e.InvoiceID != "" {
		m["invoiceId"] = e.InvoiceID
	}
	if e.CustomerID != "" {
		m["customerId"] = e.CustomerID
	}
	if e.Channel != "" {
		m["channel"] = e.Channel
	}
	if e.Notes != "" {
		m["notes"] = e.Notes
	}
	b, err := json.Marshal(m)
	if err != nil {
		return "{}"
	}
	return string(b)
}

// Sha256Hex returns the hex SHA-256 digest of input.
func Sha256Hex(input string) string {
	h := sha256.Sum256([]byte(input))
	return hex.EncodeToString(h[:])
}

// ComputeHash hashes an entry given its prevHash.
func ComputeHash(prevHash string, e *AuditEntry) string {
	return Sha256Hex(prevHash + StablePayload(e))
}

// AppendAuditEntry returns the log with the new hashed entry appended.
// Pure: does NOT mutate the input slice.
func AppendAuditEntry(log []AuditEntry, newEntryData AuditEntry) []AuditEntry {
	prevHash := GenesisHash
	if len(log) > 0 {
		prevHash = log[len(log)-1].Hash
	}
	entry := newEntryData
	entry.PrevHash = prevHash
	entry.Hash = ComputeHash(prevHash, &newEntryData)
	out := make([]AuditEntry, len(log)+1)
	copy(out, log)
	out[len(log)] = entry
	return out
}

// ChainVerification reports chain validity.
type ChainVerification struct {
	Valid         bool
	BrokenAtIndex int // -1 when valid
	Entries       int
}

func NewChainVerification(valid bool, brokenAtIndex int, entries int) ChainVerification {
	return ChainVerification{Valid: valid, BrokenAtIndex: brokenAtIndex, Entries: entries}
}

// VerifyChain recomputes the entire chain from scratch and returns the first
// broken index (-1 if the whole chain verifies).
func VerifyChain(log []AuditEntry) ChainVerification {
	for i := range log {
		entry := log[i]
		expectedPrev := GenesisHash
		if i > 0 {
			expectedPrev = log[i-1].Hash
		}
		if entry.PrevHash != expectedPrev {
			return NewChainVerification(false, i, len(log))
		}
		recomputed := ComputeHash(entry.PrevHash, &entry)
		if entry.Hash != recomputed {
			return NewChainVerification(false, i, len(log))
		}
	}
	return NewChainVerification(true, -1, len(log))
}
